README.md
# User API web application

It is a basic NodeJS web application exposing REST API that creates and stores user parameters in [Redis database](https://redis.io/).

## Functionality

1. Start a web server
2. Create a user

## Installation

This application is written on NodeJS and it uses Redis database.

1. [Install NodeJS](https://nodejs.org/en/download/)

2. [Install Redis](https://redis.io/download)

3. Install application

Go to the root directory of the application (where `package.json` file located) and run:

```
npm install 
```

## Usage

1. Start a web server

From the root directory of the project run:

```
npm start
```

It will start a web server available in your browser at http://localhost:3000.

2. Create a user

Send a POST (REST protocol) request using terminal:

```bash
curl --header "Content-Type: application/json" \
  --request POST \
  --data '{"username":"sergkudinov","firstname":"sergei","lastname":"kudinov"}' \
  http://localhost:3000/user
```

It will output:

```
{"status":"success","msg":"OK"}
```

Another way to test your REST API is to use [Postman](https://www.postman.com/).

## Testing

From the root directory of the project, run:

```
npm test
```

## Authors

Nicolas Saint
Thomas Chopin


CHANGELOG.md
# Changelog

## v1.0.0

- Create user functionality
- Create web server


test/user.router.js
const app = require("../src/index");
const chai = require("chai");
const chaiHttp = require("chai-http");
const userController = require("../src/controllers/user");
const db = require("../src/dbClient");

chai.use(chaiHttp);

describe("User REST API", () => {
  beforeEach(() => {
    // Clean DB before each test
    db.flushdb();
  });

  after(() => {
    app.close();
    db.quit();
  });

  describe("POST /user", () => {
    it("create a new user", (done) => {
      const user = {
        username: "sergkudinov", 
        firstname: "Sergei",
        lastname: "Kudinov",
      };
      chai
        .request(app)
        .post("/user")
        .send(user)
        .then((res) => {
          chai.expect(res).to.have.status(201);
          chai.expect(res.body.status).to.equal("success");
          chai.expect(res).to.be.json;
          done();
        })
        .catch((err) => {
          throw err;
        });
    });

    it("pass wrong parameters", (done) => {
      const user = {
        firstname: "Sergei",
        lastname: "Kudinov",
      };
      chai
        .request(app)
        .post("/user")
        .send(user)
        .then((res) => {
          chai.expect(res).to.have.status(400);
          chai.expect(res.body.status).to.equal("error");
          chai.expect(res).to.be.json;
          done();
        })
        .catch((err) => {
          throw err;
        });
    });
  });

  describe("GET /user", () => {
    // TODO Create test for the get method
    it("create a new user", (done) => {
      const user = {
        username: "sergkudinov",
        firstname: "Sergei",
        lastname: "Kudinov",
      };

      userController.create(user, () => {
        chai
          .request(app)
          .get("/user/"+user.username)
          .then((res) => {
            chai.expect(res).to.have.status(200);
            chai.expect(res.body.status).to.equal("Succès");
            chai.expect(res).to.be.json;
            done();
          })
          .catch((err) => {
            throw err;
          });
      });
    });

    it("pass wrong parameters", (done) => {
      const user = {
        firstname: "Sergei",
        lastname: "Kudinov",
      };
      userController.create(user, () => {
        chai
          .request(app)
          .get("/user/"+user.username)
          .then((res) => {
            chai.expect(res).to.have.status(404);
            chai.expect(res.body.status).to.equal("erreur");
            chai.expect(res).to.be.json;
            done();
          })
          .catch((err) => {
            throw err;
          });
      });
    });
  });
});


test/user.controller.js
const { expect } = require("chai");
const userController = require("../src/controllers/user");
const db = require("../src/dbClient");

describe("User", () => {
  beforeEach(() => {
    // Clean DB before each test
    db.flushdb();
  });

  describe("Create", () => {
    it("create a new user", (done) => {
      const user = {
        username: "sergkudinov",
        firstname: "Sergei",
        lastname: "Kudinov",
      };
      userController.create(user, (err, result) => {
        expect(err).to.be.equal(null);
        expect(result).to.be.equal("OK");//test
        done();
      });
    });

    it("passing wrong user parameters", (done) => {
      const user = {
        firstname: "Sergei",
        lastname: "Kudinov",
      };
      userController.create(user, (err, result) => {
        expect(err).to.not.be.equal(null);
        expect(result).to.be.equal(null);
        done();
      });
    });

    it("avoid creating an existing user", (done) => {
      //   // TODO create this test
      //   // Warning: the user already exists

      const user = {
        username: "lilparenzi",
        firstname: "Léo",
        lastname: "Parrent",
      };
      userController.create(user, () => {

        userController.create(user, (err,result) => {

        expect(err).to.not.be.equal(null);
        expect(result).to.be.equal(null);
        done();
        })
      });

    
    });
  });

  // TODO Create test for the get method
  describe("Get", () => {
    it("get a user by username", (done) => {
      //1. First, create a user to make this unit test independent from the others
      const user = {
        username: "tomchopin",
        firstname: "Thomas",
        lastname: "Chopin",
      };

      userController.create(user, () => {
        userController.get(user.username, (err, result) => {
          expect(err).to.be.equal(null);
          expect(result).to.be.deep.equal({
            firstname: "Thomas",
            lastname: "Chopin",
          });
          done();
        });
      });
    });
    

    it("cannot get a user when it does not exist", (done) => {
      // Check si l'utilisateur est valide
      userController.get("invalid", (err, result) => {
        expect(err).to.not.be.equal(null);
        expect(result).to.be.equal(null);
      });
      done();
    });
  });
});

src/routes/user.js
const express = require("express");
const userController = require("../controllers/user");

const userRouter = express.Router();

userRouter
  .post("/", (req, resp) => {
    userController.create(req.body, (err, res) => {
      let respObj;
      if (err) {
        respObj = {
          status: "error",
          msg: err.message,
        };
        return resp.status(400).json(respObj);
      }
      respObj = {
        status: "success",
        msg: res,
      };
      resp.status(201).json(respObj);
    });
  })
  .get("/:username", (req, resp, next) => {
    // Express URL params - https://expressjs.com/en/guide/routing.html
    //TODO Create get method API
    const username = req.params.username;

    userController.get(username, (err, res) => {
      let reponse;

      if (err) {
        reponse = {
          status: "erreur",
          message: err.message,
        };
        resp.status(404).json(reponse);
      } else {
        reponse = {
          status: "Succès",
          message: res,
        };
        resp.status(200).json(reponse);
      }
    });
    next();
  });

module.exports = userRouter;
